keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,980 @@
+
+```json
+"ClusterCIDRConfig": {
+  "Metadata": {
+    "finalizers": [
+      "networking.kubernetes.io/cluster-cidr-config-finalizer",
+    ]
+  },
+  "Status": {
+    "Conditions": [
+      {
+        "Type": "active",
+        "Status": "False"
+      },
+      {
+        "Type": "terminating",
+        "Status": "True",
+        "Reason": "cidr-in-use",
+        "Message": "ClusterCIDRRange in use by nodes: ['node1, 'node2']"
+      }
+    ]
+  }
+}
+```
+
+#### Example: Allocations
+
+```go
+[
+  {
+      // For existing clusters this is the same as ClusterCIDR
+      CIDR:            "10.0.0.0/8",
+      // Default for nodes not matching any other rule
+      NodeSelector:    nil,
+      // For existing API this is the same as NodeCIDRMaskSize
+      PerNodeMaskSize: 24,
+  },
+  {
+      CIDR:            "172.16.0.0/14",
+      // Another range, also allocate-able to any node
+      NodeSelector:    nil,
+      PerNodeMaskSize: 24,
+  },
+  {
+      CIDR:            "10.0.0.0/8",
+      NodeSelector:    { key: "np" op: "IN" value:["np1"] },
+      PerNodeMaskSize: 26,
+  },
+  {
+      CIDR:            "192.168.0.0/16",
+      NodeSelector:    { key: "np" op: "IN" value:["np2"] },
+      PerNodeMaskSize: 26,
+  },
+  {
+      CIDR:            "5.2.0.0/16",
+      NodeSelector:    { "np": "np3" },
+      PerNodeMaskSize: 20,
+  },
+  {
+      CIDR:            "fd12:3456:789a:1::/64",
+      NodeSelector:    { "np": "np3" },
+      PerNodeMaskSize: 112,
+  },
+  ...
+]
+```
+
+Given the above config, a valid potential configuration might be:
+
+```
+{"np": "np1"} --> "10.0.0.0/26"
+{"np": "np2"} --> "192.16.0.0/26"
+{"np": "np3"} --> "5.2.0.0/20", "fd12:3456:789a:1::/112"
+{"np": "np4"} --> "172.16.0.0/24"
+```
+
+### Controller
+
+Implement a new
+[NodeIPAM controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodeipam)
+The controller will set up watchers on the `ClusterCIDRConfig` objects and the
+`Node` objects.
+
+This controller relies on being a single writer (just as the current NodeIPAM
+controller does as well). In the case of HA control planes with multiple
+replicas, there will have to be some form of leader election to enforce only 1
+active leader. This KEP proposes re-using the kube-controller-manager leader
+election to pick a active controller.
+
+#### Data Structures
+
+We will use maps to store the allocated ranges and which node is using the
+range. Because the number of nodes is expected to be on the order of thousands,
+more sophisticated data structures are likely not required.
+
+Prior investgations [here](https://github.com/kubernetes/kubernetes/pull/90184)
+suggest that maps storing allocations will perform well under the number of
+nodes we expect.
+
+#### Startup
+
+-   Fetch list of `ClusterCIDRConfig` and build internal data structure
+-   If they are set, read the `--cluster-cidr` and `--node-cidr-mask-size` flags
```

Comment:
Under what conditions do we create this?  Always?  Only if there are no other ranges defined?  Only if there is no range that exactly matches the flags?

Reply:
If we are using the new node IPAM controller, we should always create this backend data structure. Is there a disadvantage to doing so?